aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -173,6 +173,33 @@ TORCH_META_FUNC(scatter_add)
   scatter_meta_impl(*this, self, dim, index, src, "add");
 }
 
+TORCH_PRECOMPUTE_META_FUNC(index_add)
+(const Tensor& self, int64_t dim, const Tensor& index, const Tensor& source, const Scalar& alpha) {
```

Comment:
I just realized before landing this - that check that TensorIterator is handling only runs in the impl function (because that's where we construct the TensorIterator). That means that the meta function won't catch everything:

```
>>> a = torch.ones(5, 5, 5, 5)
>>> b = torch.ones(5, 5, 8, 5)
>>> i = torch.ones(5, dtype=torch.int)
# This fails properly
>>> torch.index_add(a, 0, i, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The size of tensor a (5) must match the size of tensor b (8) at non-singleton dimension 1

>>> a = torch.ones(5, 5, 5, 5, device='meta')
>>> b = torch.ones(5, 5, 8, 5, device='meta')
>>> i = torch.ones(5, dtype=torch.int, device='meta')
# But this passes?
>>> torch.index_add(a, 0, i, b)
tensor(..., device='meta', size=(5, 5, 5, 5))
```

So... we need some way to run those TensorIterator checks in the meta function? That looks difficult because the inputs to `borrowing_binary_op` aren't the original inputs: they're slices of the inputs. Some options that I can think of:
- (1) try to manually emulate the checks that TI runs in this constructor. We run the risk of missing some steps that TI does though. Also, it was mentioned in this thread that we hit some weird vmap errors originally trying to do this.
- (2) try to get `index_add` to inherit from `TensorIteratorBase`. This... might work, but might be more involved / you'd have to try it out. We'd need to create the intermediates that you pass into `build_borrowing_binary_op` directly in the meta function.
- (3) Throw something like "if device == kMeta, `TensorIteratorBase::borrowing_binary_op`"  into the meta function.

What do you guys think? (2) Sounds maybe nicer, but I'm not sure if you'll run into any problems trying to do it. (3) sounds hackier but easier to add in, to correct the meta function.

Reply:
@bdhirsh - Thanks for pointing this out. Sorry for missing this earlier.

2nd option does sound better and less of a hack, but I temporarily went for the 3rd method. It does fix things, and I added a comment there for reference. I'm running a little tight on other PRs, hence couldn't try the 2nd method - but I'll give it a try this week to see if it works.

I'm sorry for the delay, but I'll try getting back on the 2nd option by the end of this week, but if you think it's okay to go for the 3rd option (and the current state of this PR) - that's also okay to me, I'll try the 2nd option and if that fixes without any failures/problems - I can raise a separate PR.

Reply:
I'll try to get this landed and we can leave that as a followup :) LGTM

Reply:
BTW the hack here is incorrect when the input is scalar

```
IndexError: select() cannot be applied to a 0-dim tensor. 
```